### PR TITLE
WIP: Refactor of cloud game overlord and overworker

### LINKED ISSUE #68: Proposal: Split binaries for overlordhost and overlord into separate binary
Currently, cloud-game generates only one binary for both the program included in a single binary. While running we have to explicitly set flags to indicate which mode we intend it to run.
I am suggesting to split the two binaries into two different executables. They might share some of the same library packages but should live in different executables. I think there are some benefits - like - 
```
1. we could include runtime flags specific to one program,
2. We could do different versioning for overlordhost and overlord,
3. one executable doesn't include the life cycle of the other one.
```

Here is some package restructure proposal that I think would suit nicely
```
cloud-game
|-- assets     // includes all static .so or game file
|-- assests/games
|-- assests/lib

|-- cmd         // will include all the executables
|-- cmd/overlord  // will include the main.go for overlord. can use spf13/cobra
|-- cmd/overlordhost  // main.go for overlordhost

|-- docs      // includes all the static docs
|-- docs/design
|-- docs/userguide
|-- docs/site    // includes static site pages for projects

|-- hack       // Includes various scripts, build files
|-- hack/docker // includes docker build files
|-- hack/config  // includes various config files like prometheus.conf
|-- hack/k8s    // includes k8s runtime files
|-- hac/scripts  // various scripts to build env or others

|-- pkg  // includes all go/ cgo librarie codes. ie-
|-- pkg/api
|-- pkg/util
|-- pkg/config
|-- pkg/webrtc

| -- static   // include project-level static resources like the logo
| -- static/logo
| -- static

|-- web      // all the files for the web frontend
|-- web/

|-- vendor // go vendor directory

|-- Makefiles
|-- go.mod
|-- go.sum
|-- README.md
```

@giongto35  Let me know what you think about this? I can go forward with a PR if it seems okay. Thanks for this awesome project. 😄

### PR DESCRIPTION
**Pull request by sadlil**
Work In Progress.

Fixes #68

### CODE REVIEW

**Comment by giongto35** [comment]
@_@ wow, it's a massive change. When it finishes, it gonna be awesome. (y)

**Comment by sadlil** [comment]
@giongto35 Hope so. I am still doing the changes. Hopefully, you can start to review some part from now. :) Thanks, man.

**Comment by giongto35** [comment]
Yes I'm checking :P

File: pkg/worker/handlers.go [hunk 1]
```diff
@@ -24,23 +24,23 @@ type Handler struct {
 	// Raw address of overlord
 	overlordHost string
 	// Rooms map : RoomID -> Room
-	rooms map[string]*room.Room
+	rooms map[string]*room2.Room
```
Comment (giongto35) [inline pkg/worker/handlers.go]: Why do you rename it to room2 and storage2?

Comment (sadlil) [reply pkg/worker/handlers.go]: Ah. None. It was my refactoring tool. Will fix it. Thanks.

Comment (giongto35) [reply pkg/worker/handlers.go]: What is your tool for lint. It's so magic. Teach me senpai.

**Comment by giongto35** [comment]
I have a new update on encoder. You may need to introduce new files.
https://github.com/giongto35/cloud-game/pull/70/files

added files:
> encoder/type.go
> h264encoder/encoder.go

File: pkg/worker/handlers.go (hunk 1 above)
Comment (sadlil) [reply pkg/worker/handlers.go]: I normally use all the common go tools -
For lint and format:
```
- go fmt
- go import
- golangci-lint
```

For moving packages around I use `gomvpkg` from - https://github.com/golang/tools/tree/master/cmd

And for statci code check i use staticcheck. You will find the installations - `./hack/scripts/install_tools.sh`. And how to uses in the `Makefiles`.

**Comment by sadlil** [comment]
@giongto35 Go ahead and merge, please. I will fix the conflicts in my branch. :)

**Comment by sadlil** [comment]
@giongto35 I have fixed all the merge issues. But while running My game is failing as my `nanoarc.go:L345` is failing to find the game. I have moved the emulators to `assets/emulator/libretro/cores` directory as I think this is more appropriate palace. Can help me with this. What is this url represents? Is it relative to something? Or I need to put the .so files in the same directory?

**Comment by sadlil** [comment]
To me the problem looks like a Mac issue. Did you happen to encounter this problem before? Let me know. @giongto35

**Comment by sadlil** [comment]
@giongto35 I have completed the first version of refactoring the code.  Apart from the issue with `C.dlopen` everything else works fine. Please have a look. Help fix the error and if its looks okay - lets merge it. For reference, we need to fix the following in future iterations for refactoring

1. Use flags embedded in binary instead of global flags
2. Few more code refactoring inside overlord and worker codes
3. Rewrite Readme
4. Organize and fix docker files and builds
5. Get rid of `pkg/config` and reorganize `pkg/util`
6. Write E2e tests and move `main_test.go` to `tests/e2e` using ginkgo - example - https://github.com/appscode/voyager/blob/master/test/e2e/e2e.go

**Comment by giongto35** [comment]
Sorry @sadlil, I have a problem of github notification, so I forgot to check. :D
This is massive and necessary. I 'm not aware of the problem with Mac because I develop it on Ubuntu. You re so good figuring it out.
And I prioritize your merge, Hopefully to get it merged today or tomorrow. Will not spoil any conflict :p

**Comment by sadlil** [comment]
@giongto35 It okay. No worries. Take your time to review and please have a short testing session to check everything works well with this patch.

**Comment by giongto35** [comment]
FYI, vendor removed all C files related to x264 :( . I need to use https://github.com/goware/modvendor to get it back.
BTW, I saw you using vendor but don't have any script based on vendor. I think vendor can only help when we build with `-mod=vendor`

**Comment by giongto35** [comment]
I love the new log <3

**Pull request merged**